041-opa-integration/proposal.md

```diff
@@ -0,0 +1,149 @@
+# Summary
+
+This proposal outlines the beginnings of support for policy enforcement. OPA 
+(Open Policy Agent) will be used as the first policy manager to integrate with.
+
+
+# Motivation
+
+Generally speaking, anything that exposes an HTTP API (whether an individual 
+micro-service or an application as a whole) needs to control who can run those 
+APIs and when. 
+
+Concourse, as a CI platform, also needs to control who can do something and when,
+which can be done by integrating with a policy manager. Before any action, send
+a policy check request, with data that describes the action, to the policy manager,
+and only continue if the policy manager replies a "pass".
+
+Possible policies to apply could be:
+
+* Steps, e.g. code security scan, by certain resource types must be done.
+* Some team's tasks must be run with certain tag.
+* Some docker image registries are not allowed.
+* Some docker images are not not allowed.
+* and so on ...
+
+
+# Proposal
+
+## Support multiple policy managers
+
+Like how multiple credential managers are supported, Concourse should allow other
+policy managers than OPA. Thus an interface of `PolicyCheck` should be defined, and
+OPA is one of the implementations.
+
+## Policy check points
+
+* All API calls, for example `set-pipeline`, will also go through policy checks. If
+the check doesn't pass, API should return HTTP code 403 (forbidden).
+* `UsingImage` action will be sent to OPA before Concourse launches a container in
```

Comment:
I can also imagine wanting to prevent the use of `privileged: true` for tasks (perhaps even on a per-team basis), which also feels like it should be handled as some `PolicyCheck` (and probably OPA?).

This makes me wonder: are there any other things besides images (and `privileged`) that operators would be interested in checking against OPA? The answer may just be no, but for arguments sake, what if Concourse didn't try to be opinionated about what actions to check against, and instead supported 4 actions beyond to the API actions: `Check`, `Get`, `Put`, and `Task`. Each action's `data` was the entire step definition.

Granted, this would mean that if all you cared about was the image used, you'd need 4 identical policy check definitions - but, by it's nature, this also affords operators more flexibility in what kind of checks they perform. This may be overkill depending on the use cases in practice, though. However, it does feel more similar to the approach taken for API endpoints (being unopinionated and sending the entire body of the API call).